eth/api.go

```diff
@@ -351,10 +352,39 @@ func (api *PrivateDebugAPI) Preimage(ctx context.Context, hash common.Hash) (hex
 	return nil, errors.New("unknown preimage")
 }
 
+// BadBlockArgs represents the entries in the list returned when bad blocks are queried.
+type BadBlockArgs struct {
+	Hash  common.Hash            `json:"hash"`
+	Block map[string]interface{} `json:"block"`
+	RLP   string                 `json:"rlp"`
+}
+
 // GetBadBLocks returns a list of the last 'bad blocks' that the client has seen on the network
 // and returns them as a JSON list of block-hashes
-func (api *PrivateDebugAPI) GetBadBlocks(ctx context.Context) ([]core.BadBlockArgs, error) {
-	return api.eth.BlockChain().BadBlocks()
+func (api *PrivateDebugAPI) GetBadBlocks(ctx context.Context) ([]BadBlockArgs, error) {
+	blocks := api.eth.BlockChain().BadBlocks()
+	responseBlocks := make([]BadBlockArgs, 0, len(blocks))
+	for _, block := range blocks {
+		rlpData, err := rlp.EncodeToBytes(block)
+		if err != nil {
+			return responseBlocks, err
+		}
+		var (
+			blockdata map[string]interface{}
+			berr      error
+		)
+		// A bad block may be so malformed that it the expanded format fails to generate
+		// correctly. If so, provide the caller with error message and RLP anyway
+		if blockdata, berr = ethapi.RpcMarshalBlock(block, true, true); berr != nil {
+			blockdata["error"] = berr.Error()
+		}
+		responseBlocks = append(responseBlocks, BadBlockArgs{
+			Hash:  block.Hash(),
+			Block: blockdata,
+			RLP:   fmt.Sprintf("%x", rlpData),
+		})
+	}
```

Comment:
This might be a bit cleaner as:

```go
	blocks := api.eth.BlockChain().BadBlocks()
	results := make([]*BadBlockArgs, len(blocks))

	var err error
	for i, block := range blocks {
		results[i] = &BadBlockArgs{
			Hash:  block.Hash(),
		})
		if results[i].RLP, err = rlp.EncodeToBytes(block); err != nil {
			results[i].RLP = err.Error() // Hacky, but hey, it works
		}
		if results[i].Block, err = ethapi.RPCMarshalBlock(block, true, true); err != nil {
			results[i].Block = map[string]interface{}{"error": berr.Error()}
		}
	}
	return results, nil
```

Note, your code would have crashed on marshal errors because the `blockdata` map was not created.